File: leetcode_problems/typescript/src/lc144_binary_tree_preorder_traversal.ts
function preorderTraversalRecursion(node: TreeNode | null, res: number[]) {
  if (node === null) {
    return;
  }

  res.push(node.val);
  preorderTraversalRecursion(node.left, res);
  preorderTraversalRecursion(node.right, res);
}

function preorderTraversal(root: TreeNode | null): number[] {
  let res: number[] = [];
  preorderTraversalRecursion(root, res);
  return res;
}

File: leetcode_problems/cpp/lc144_binary_tree_preorder_traversal.cpp
#include <vector>
using std::vector;

struct TreeNode
{
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution
{
public:
    void traverse(TreeNode *node, vector<int> &res)
    {
        if (node == nullptr)
        {
            return;
        }

        res.push_back(node->val);
        traverse(node->left, res);
        traverse(node->right, res);
    }

    vector<int> preorderTraversal(TreeNode *root)
    {
        vector<int> res;
        traverse(root, res);
        return res;
    }
};
